File: Winter2022/examples/java/Lecture 9/CanvasDemo/app/src/main/java/mobiledev/unb/ca/canvasdemo/BubbleView.java
package mobiledev.unb.ca.canvasdemo;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.View;

import java.util.Random;

@SuppressLint("ViewConstructor")
public class BubbleView extends View {
    private static final String TAG = "BubbleView";
    private static final int BITMAP_SIZE = 64;
    private static final int MAX_STEP = 10;

    // Display dimensions
    private final int mDisplayWidth;
    private final int mDisplayHeight;

    final private Bitmap mBitmap;

    private float mCurrX;
    private float mCurrY;
    private int mStepX;
    private int mStepY;
    private final Handler h;

    final private Paint mPainter = new Paint();

    public BubbleView(Context context, Bitmap bitmap, int displayWidth, int displayHeight) {
        super(context);

        h = new Handler(Looper.getMainLooper());
        this.mBitmap = Bitmap.createScaledBitmap(bitmap,
                BITMAP_SIZE, BITMAP_SIZE, false);

        mDisplayWidth = displayWidth;
        mDisplayHeight = displayHeight;

        // Start in roughly the centre
        mCurrX = mDisplayWidth / 2.0f;
        mCurrY = mDisplayHeight / 2.0f;

        Random r = new Random();
        // Pick a random x and y step
        mStepX = r.nextInt(MAX_STEP) + 1;
        mStepY = r.nextInt(MAX_STEP) + 1;
        // Pick a random x and y direction to step
        mStepX = r.nextBoolean() ? mStepX : -mStepX;
        mStepY = r.nextBoolean() ? mStepY : -mStepY;

        mPainter.setAntiAlias(true);
    }

    protected void onDraw(Canvas canvas) {
        if (move()) {
            canvas.drawBitmap(mBitmap, mCurrX, mCurrY, mPainter);
            h.postDelayed(r, 30);
        }
    }

    private final Runnable r = () -> {
        invalidate();
        try {
            Thread.sleep(20);
        } catch (InterruptedException e) {
            Log.i(TAG, "InterruptedException");
        }
    };

    private boolean move() {
        mCurrX = mCurrX + mStepX;
        mCurrY = mCurrY + mStepY;

        // Return true if the BubbleView is on the screen
        return mCurrX <= mDisplayWidth &&
                mCurrX + BITMAP_SIZE >= 0 &&
                mCurrY <= mDisplayHeight &&
                mCurrY + BITMAP_SIZE >= 0;
    }
}

File: Fall2022/examples/Lecture 9/CanvasDemo/app/src/main/java/mobiledev/unb/ca/canvasdemo/BubbleView.kt
package mobiledev.unb.ca.canvasdemo

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Paint
import android.view.View
import java.util.Random
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

@SuppressLint("ViewConstructor")
class BubbleView (context: Context?, displayWidth: Int, displayHeight: Int) :
    View(context) {

    // Reference to the thread job
    private var mMoverFuture: ScheduledFuture<*>? = null

    // Display dimensions
    private var mStepX: Int
    private var mStepY: Int
    private val mPainter = Paint()

    private val mDisplayWidth: Int = displayWidth
    private val mDisplayHeight: Int = displayHeight
    private var mCurrX: Float = displayWidth / 2.0f
    private var mCurrY: Float = displayHeight / 2.0f

    // Reference to the scaled bitmap object
    private var scaledBitmap: Bitmap

    @Synchronized
    override fun onDraw(canvas: Canvas) {
        canvas.drawBitmap(scaledBitmap, mCurrX, mCurrY, mPainter)
    }

    fun startMovement() {
        // Creates a WorkerThread
        val executor = Executors.newScheduledThreadPool(1)

        // Execute the run() in Worker Thread every REFRESH_RATE milliseconds
        // Save reference to this job in mMoverFuture
        mMoverFuture = executor.scheduleWithFixedDelay({
            val stillOnScreen = moveWhileOnScreen()
            if (stillOnScreen) {
                this@BubbleView.postInvalidate()
            }
        }, 0, REFRESH_RATE.toLong(), TimeUnit.MILLISECONDS)
    }

    private fun moveWhileOnScreen(): Boolean {
        mCurrX += mStepX
        mCurrY += mStepY

        // Return true if the BubbleView is on the screen
        return mCurrX <= mDisplayWidth
                && mCurrX + BITMAP_SIZE >= 0
                && mCurrY <= mDisplayHeight
                && mCurrY + BITMAP_SIZE >= 0
    }

    companion object {
        private const val BITMAP_SIZE = 64
        private const val MAX_STEP = 10
        private const val REFRESH_RATE = 40
    }

    init {
        // Set the bubble image from the drawable resource
        val bitmap = BitmapFactory.decodeResource(resources, R.drawable.b64)
        scaledBitmap = Bitmap.createScaledBitmap(bitmap, BITMAP_SIZE, BITMAP_SIZE,false)

        // Start in roughly the centre
        val r = Random()

        // Pick a random x and y step
        mStepX = r.nextInt(MAX_STEP) + 1
        mStepY = r.nextInt(MAX_STEP) + 1

        // Pick a random x and y direction to step
        mStepX = if (r.nextBoolean()) mStepX else -mStepX
        mStepY = if (r.nextBoolean()) mStepY else -mStepY

        // Smooth out the edges
        mPainter.isAntiAlias = true
    }
}